### PR TITLE
Use dictionary lookup while resolving assembly references

### LINKED ISSUE #66410: High CPU Utilization with Very Basic Keypresses
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/High-CPU-Utilization-with-Very-Basic-Key/10247502)._

Internal ticket: https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1714591

---
[severity:It's more difficult to complete my work] [regression] [worked-in:17.4]
Please see attached.  I am seeing high CPU utilization that is occurring in a basic .cs file.  Upwards to 50% with delays/hitches on the UI thread.

This seems new for 17.5

---
#### Feedback Bot on 1/8/2023, 06:56 PM: 

(private comment, text removed)
#### Feedback Bot on 1/10/2023, 08:51 AM: 

(private comment, text removed)
#### Cyrus Najmabadi [MSFT] on 1/11/2023, 00:26 PM: 

(private comment, text removed)
#### Mike-E-angelo on 1/11/2023, 01:16 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #66410.

### CODE REVIEW

**Comment by Xyncgas** [comment]
have you done any benchmarks for how much this would be better

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs [hunk 1]
```diff
@@ -310,6 +319,15 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
                 // Rebind assembly references that were initially missing. All bindings established above
                 // are against explicitly specified references.
 
+                // We only need to resolve against implicitly resolved assemblies,
+                // since we already resolved against explicitly specified ones.
+                var implicitAssembliesMap = new MultiDictionary<string, (AssemblyData DefinitionData, int DefinitionIndex)>(AssemblyIdentityComparer.SimpleNameComparer);
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs:324]: >new MultiDictionary [](http://example.com/codeflow?start=44&length=19)

Can we also pass `capacity: implicitAssemblies.Count`? Similarly for `explicitAssembliesMap` above.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**